src/Workspaces/Core/Portable/Execution/Serialization/Serializer_Asset.cs

```diff
@@ -0,0 +1,287 @@
+
+            var name = reader.ReadString();
+            var folders = reader.ReadArray<string>();
+            var sourceCodeKind = reader.ReadInt32();
+            var filePath = reader.ReadString();
+            var isGenerated = reader.ReadBoolean();
+
+            return new DocumentChecksumObjectInfo(documentId, name, folders, (SourceCodeKind)sourceCodeKind, filePath, isGenerated);
+        }
+
+        public void Serialize(ITemporaryStorageWithName storage, SourceText text, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            writer.WriteInt32((int)text.ChecksumAlgorithm);
+            writer.WriteString(text.Encoding?.WebName);
+
+            // TODO: refactor this part in its own abstraction (Bits) that has multiple sub types
+            //       rather than using enums
+            if (storage != null && storage.Name != null)
+            {
+                writer.WriteInt32((int)SerializationKinds.MemoryMapFile);
+                writer.WriteString(storage.Name);
+                writer.WriteInt64(storage.Size);
+                return;
+            }
+
+            writer.WriteInt32((int)SerializationKinds.Bits);
+            writer.WriteString(text.ToString());
+        }
+
+        private SourceText DeserializeSourceText(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // REVIEW: why IDE services doesnt care about checksumAlgorithm?
+            var checksumAlgorithm = (SourceHashAlgorithm)reader.ReadInt32();
+            var webName = reader.ReadString();
+            var encoding = webName == null ? null : Encoding.GetEncoding(webName);
+
+            var kind = (SerializationKinds)reader.ReadInt32();
+            if (kind == SerializationKinds.MemoryMapFile)
+            {
+                var name = reader.ReadString();
+                var size = reader.ReadInt64();
+
+                var tempService = _workspaceServices.GetService<ITemporaryStorageService>() as ITemporaryStorageService2;
+                var storage = tempService.AttachTemporaryTextStorage(name, size, encoding, cancellationToken);
+
+                return storage.ReadText(cancellationToken);
+            }
+
+            // TODO: should include version info here as well?
+
+            var textService = _workspaceServices.GetService<ITextFactoryService>();
+            using (var textReader = new StringReader(reader.ReadString()))
+            {
+                return textService.CreateText(textReader, encoding, cancellationToken);
+            }
+        }
+
+        public void Serialize(string language, CompilationOptions options, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // TODO: language specific, should I put this ability in compilation layer?
```

Comment:
opened a bug to compiler team. problem is Options are not pure data. it has func as well such as resolvers and such.

anyway, opened a bug to them.